### PR TITLE
Use mocha to determine what files to run

### PR DESCRIPTION
**Pull request by gustavnikolaj**

### CODE REVIEW

**Comment by alexjeffburke** [comment]
Looks like a really nice change to generate the list of files first and hand them off.

One thing I notice, this seems to at least partially reimplement the way mocha finds files. I guess I deal would be to `require` something from within the 'library' portion of mocha but there's a high probability I'm preaching to the converted and that interface just doesn't exist yet :)

File: lib/findFilesMochaWouldRun.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+const { resolve } = require("path");
+const minimatch = require("minimatch");
+// TODO: Mocha should be a peerDep
+// TODO: Find out which version of mocha introduced the lookupFiles method.
+const utils = require("mocha/lib/utils");
+
+/* This file is lifted out of `bin/_mocha` [1]. We duplicate mocha's logic to
+ * find the relevant files to build our dependency graph from.
+ * [1]: https://github.com/mochajs/mocha/blob/d401ee198bd5026bc61641ad26eab31e6746713f/bin/_mocha
+ */
```
Comment (gustavnikolaj) [inline lib/findFilesMochaWouldRun.js]: @alexjeffburke This code is as the comment says pulled out of `bin/_mocha`. I can't require that file or other wise get to it.

But notice on the line above where I include the utils file from mocha. It is actually that code that goes and "globs" for the test files.

File: lib/loadMochaOptions.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ * Parse list.
+ */
+const list = str => str.split(/ *, */);
+
+/**
+ * Parse multiple flag.
+ */
+const collect = (val, memo) => memo.concat(val);
+
+module.exports = function loadMochaOptions(args) {
```
Comment (gustavnikolaj) [inline lib/loadMochaOptions.js]: @alexjeffburke This is also lifted out of `bin/_mocha`. I think the most elegant way is to duplicate the necessary bits at first. If we are hugely successful, we might be able to convince mocha-core to pull it out into requireable methods.

File: lib/loadMochaOptsIntoArgs.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  // that is found in a related mocha.opts file. The mocha.opts file may be
+  // referenced in the arguments present in progress.argv so in order to avoid
+  // extending it twice the environment variable is set.
+
+  if (!process.env.LOADED_MOCHA_OPTS) {
+    getOptions();
+  }
+
+  return process.argv;
+};
```
Comment (gustavnikolaj) [inline lib/loadMochaOptsIntoArgs.js]: @alexjeffburke this file is just calling into mocha's options loader. The function is here to make sure that we don't need to pollute the rest of the project with intimate knowledge about mocha's internals (at least not regarding this) - mocha's `getOptions` is not safe against being called more than once and relies on the caller checking the presence of the environment variable.

**Comment by gustavnikolaj** [comment]
Haha @alexjeffburke  - no worries. I thought it was a little bit unfair critique...

I missed your update though, my browser tab loaded in your new comment but not your edit. So I'm sorry about the extra pings in the comment. Those are my replies to the initial unedited comment.

**Comment by alexjeffburke** [comment]
@gustavnikolaj hahaha snap, apologies for not reading the diff with my brain fully active!

**Pull request merged**